### PR TITLE
fix(lspinfo): improve lspinfo behavior

### PR DESCRIPTION
**Pull request by glepnir**
problem: if file is a library file it will reuse a client .  now lspinfo will it running in single mode

solution:  add `single_buffers` in client object if file is running in single file mode passed it to this table. when show lspinfo can use this field to check file  run in workspace or single file mode.

### CODE REVIEW

File: lua/lspconfig/util.lua [hunk 1]
```diff
@@ -347,6 +359,7 @@ function M.server_per_root_dir_manager(make_config)
         return
       end
       attach_and_cache(root_dir, client_id)
+      register_client_single(vim.lsp.get_client_by_id(client_id))
```
Comment (justinmk) [inline lua/lspconfig/util.lua:362]: does this affect core lspconfig functionality, not only "lspinfo" ?

Comment (glepnir) [reply lua/lspconfig/util.lua:362]: It shouldn't have any side effects. When there is a file opened in single file mode. It create a  field called`single_buffers`  in client table. And insert this buffer. Clean up this field when the buffer is detach.

Comment (justinmk) [reply lua/lspconfig/util.lua]: but isn't it strange that LspInfo features require this kind of change? This is a hint that the `server_per_root_dir_manager` and all that "cache" code is doing too much. Why is lspconfig maintaining its own client cache...?

Comment (glepnir) [reply lua/lspconfig/util.lua]: we have a client cache but is simple cached the client id

**Comment by justinmk** [comment]
Needs a rebase since recent changes. See also https://github.com/neovim/nvim-lspconfig/issues/3376

**Pull request closed without merging**